File: switching-triage-agent/src/modules/scoring.py
"""
scoring — Score = V × U × R の計算と束ね（ベンダー単位バンドル）評価。

判断根拠は direction/scoring_methodology.md に対応する。
数式を変えるときはドキュメントとここを両方更新すること。
"""
from __future__ import annotations
import math
from collections import defaultdict


# ---- 有効確度 p_i の解決 -------------------------------------------------

def effective_p(item: dict, scenario: dict) -> dict:
    """
    実行確度 p_i を解決する。
    - contestability_prob と execution_prob が両方あれば積を使う。
      なければ exec_probability 単独にフォールバック（後方互換）。
    返り値に内訳を残す（説明可能性）。

    繰り返し延長は p を割り引かない（重要）。
    延長を確度の低下として扱うと、締切が近く実行意欲だけが低い案件が
    スコア上「捨てる候補」へ沈む。しかしそれは方法論 §3.5 の意図と逆である
    ── 現場任せでは切り替わらない案件こそ、VMOが実行主導すべき対象として
    名指しされねばならない。したがって延長は独立したフラグとして扱い、
    判定ラベル側で浮上させる（escalate_of を参照）。
    """
    contest = item.get("contestability_prob")
    exec_p = item.get("execution_prob")
    ext = int(item.get("extension_count") or 0)

    if contest is not None and exec_p is not None:
        p = contest * exec_p
        split_used = True
    else:
        p = item["exec_probability"]
        contest = None
        exec_p = None
        split_used = False

    return {
        "p": max(0.0, min(1.0, p)),
        "p_contest": contest,
        "p_exec_effective": exec_p,
        "ext_count": ext,
        "ext_escalated": ext >= int(scenario.get("ext_threshold", 2)),
        "p_split_used": split_used,
    }


# ---- 単一候補の因子計算 --------------------------------------------------

def factors(item: dict, scenario: dict) -> dict:
    """1候補について net, V_raw, d_i, U, R と有効確度内訳を計算して返す。"""
    lam = scenario["lambda_risk"]
    h = scenario["lead_time_h_months"]

    pinfo = effective_p(item, scenario)
    p = pinfo["p"]

    net = item["est_return_m"] - item["switch_cost_m"]
    V_raw = p * net * (1 - lam * (1 - p))

    d_i = item["expiry_months"] - item["switch_lead_months"]

    if d_i < 0:
        U = 0.0
    else:
        U = 1.0 / (1.0 + d_i / h)

    g = min(max(d_i, 0.0) / h, 1.0)
    R = 1.0 - item["return_sigma"] * g

    return {"net": net, "V_raw": V_raw, "d_i": d_i, "U": U, "R": R,
            "p_effective": p, **pinfo}


# ---- 束ね（ベンダー単位）------------------------------------------------

def build_bundles(contracts: list[dict], scenario: dict) -> list[dict]:
    """同一ベンダーで2件以上ある契約を束ね候補として合成する。"""
    beta = scenario["beta_bundle"]
    by_vendor: dict[str, list[dict]] = defaultdict(list)
    for c in contracts:
        by_vendor[c["vendor"]].append(c)

    bundles: list[dict] = []
    for vendor, members in by_vendor.items():
        if len(members) < 2:
            continue
        n = len(members)
        ret = beta * sum(m["est_return_m"] for m in members)
        cost = sum(m["switch_cost_m"] for m in members)
        # 束ねの確度: 各メンバーの有効確度の緩めた積（オールオアナッシング性）
        prod_p = 1.0
        for m in members:
            prod_p *= effective_p(m, scenario)["p"]
        p = prod_p ** (1.0 / math.sqrt(n))
        sigma = max(m["return_sigma"] for m in members)
        expiry = min(m["expiry_months"] for m in members)
        lead = max(m["switch_lead_months"] for m in members)

        bundles.append({
            "contract_id": f"B-{vendor}",
            "contract_name": f"{vendor} 一括（{n}件）",
            "contract_type": "Bundle",
            "vendor": vendor,
            "expiry_months": expiry,
            "annual_spend_m": sum(m["annual_spend_m"] for m in members),
            "switch_lead_months": lead,
            "switch_cost_m": cost,
            "switch_cost_sigma": 0.0,
            "est_return_m": ret,
            "return_sigma": sigma,
            "exec_probability": p,          # 束ねは合成済み確度をそのまま p として持つ
            # 延長回数は構成契約の最大を引き継ぐ。1件でも延長常習なら
            # 束ねても現場任せでは動かないため、シグナルを消してはならない。
            "extension_count": max(int(m.get("extension_count") or 0) for m in members),
            "is_bundle": True,
            "members": [m["contract_id"] for m in members],
        })
    return bundles


# ---- 判定ラベル ----------------------------------------------------------

def verdict_of(d_i: float, score: float, Vn: float, U: float, R: float,
               escalated: bool = False) -> str:
    if d_i < 0:
        return "失効"
    if score >= 0.25:
        return "即着手"
    # 繰り返し延長は、確度を割り引くのではなくここで浮上させる。
    # 「現場任せでは切り替わらない」という事実は、沈めるのではなく名指しする。
    if escalated:
        return "VMO主導候補"
    if U >= 0.5 and R < 0.7:
        return "人間判断"
    if Vn > 0.3 and U < 0.4:
        return "温存"
    if Vn < 0.1:
        return "捨てる候補"
    return "検討"



# ---- メイン: 候補集合を構築してスコアリング ------------------------------

def score_all(contracts: list[dict], scenario: dict) -> dict:
    use_bundle = scenario["use_bundle"]

    # is_active=0 の契約は対象外（Single-source確定などで交渉トラックへ移したもの）
    active = [c for c in contracts if int(c.get("is_active", 1)) == 1]
    excluded = [c["contract_name"] for c in contracts if int(c.get("is_active", 1)) != 1]

    # ベースは個別契約（コピー）
    candidates = [dict(c) for c in active]
    for c in candidates:
        c.setdefault("is_bundle", False)

    bundle_decisions: list[dict] = []

    if use_bundle:
        bundles = build_bundles(active, scenario)
        bundled_ids: set = set()
        for b in bundles:
            members = [c for c in active if c["contract_id"] in b["members"]]
            fB = factors(b, scenario)
            sum_indiv_V = sum(max(factors(m, scenario)["V_raw"], 0.0) for m in members)
            adopt = max(fB["V_raw"], 0.0) >= sum_indiv_V
            bundle_decisions.append({
                "vendor": b["vendor"], "n": len(members),
                "bundle_V": round(max(fB["V_raw"], 0.0), 2),
                "indiv_V": round(sum_indiv_V, 2),
                "adopted": adopt,
            })
            if adopt:
                candidates.append(dict(b))
                bundled_ids.update(b["members"])
        candidates = [c for c in candidates
                      if not (c.get("contract_id") in bundled_ids and not c.get("is_bundle"))]

    # 因子計算
    for c in candidates:
        c.update(factors(c, scenario))

    # V の相対正規化（金額の桁で押し切らせない）
    # 下限を定数で切らない。切ると案件群の絶対金額がスコアに漏れ込み、
    # 小規模な案件群だけ一律に低スコアへ沈む（＝相対化の意図に反する）。
    # ゼロ除算だけを避ける。assets/js/scoring.js と同じ基準。
    max_V = max([max(c["V_raw"], 0.0) for c in candidates] + [1e-9])
    for c in candidates:
        c["Vn"] = max(c["V_raw"], 0.0) / max_V
        c["score"] = c["Vn"] * c["U"] * c["R"]
        c["verdict"] = verdict_of(c["d_i"], c["score"], c["Vn"], c["U"], c["R"],
                                  escalated=bool(c.get("ext_escalated")))

    candidates.sort(key=lambda x: x["score"], reverse=True)

    return {
        "as_of": scenario["as_of_date"],
        "scenario": scenario,
        "bundle_decisions": bundle_decisions,
        "excluded_inactive": excluded,
        "candidates": candidates,
    }

File: switching-triage-agent/assets/js/scoring.js
/*
 * scoring.js — Score = V × U × R の計算と束ね評価。
 *
 * src/modules/scoring.py の移植版。式・閾値は direction/scoring_methodology.md
 * に対応する。数式を変える場合は、Python側・methodology・このファイルを
 * すべて更新すること（乖離させない）。
 *
 * λ（lambdaRisk）と h（leadTimeHMonths）は FIXED_PARAMS として scenario.json
 * の値に固定し、ダッシュボードのUIでは調整できないようにしている
 * （効果の説明が難しいため。詳細は input/scenario.json 参照）。
 */

// ---- 実行確度 p の解決 -----------------------------------------------------
//
// contestProb と execProb が両方あればその積を使う（後方互換で単独値へフォールバック）。
//
// 繰り返し延長は p を割り引かない（重要）。
// 延長を確度の低下として扱うと、締切が近く実行意欲だけが低い案件がスコア上
// 「捨てる候補」へ沈む。しかしそれは方法論 §3.5 の意図と逆であり、現場任せでは
// 切り替わらない案件こそ VMO が実行主導すべき対象として名指しされねばならない。
// したがって延長は独立フラグとして扱い、判定ラベル側で浮上させる（verdictOf）。
function effectiveP(item, params) {
  const contest = item.contestProb;
  const execP = item.execProb;
  const ext = item.extCount || 0;

  let p, pContest, pExecEff, splitUsed;
  if (contest != null && execP != null) {
    pExecEff = execP;
    p = contest * execP;
    pContest = contest;
    splitUsed = true;
  } else {
    p = item.execProbability;
    pContest = null;
    pExecEff = null;
    splitUsed = false;
  }

  return {
    p: Math.max(0, Math.min(1, p)),
    pContest,
    pExecEffective: pExecEff,
    extCount: ext,
    extEscalated: ext >= params.extThreshold,
    pSplitUsed: splitUsed,
  };
}

// ---- 基準日シフトを反映した expiry (残月数) ------------------------------
//
// Python本体では as_of_date は表示ラベルのみで expiry_months は
// CSV入力時点ですでに基準日からの相対値として固定されている（動的な
// 再計算はしない）。ダッシュボードでは as_of_date を「動かして試せる」
// パラメータにする意味を持たせるため、BASE_AS_OF からのシフト分だけ
// 各契約の expiryM を前later/後ろへずらす。これは本体パイプラインには
// ない、ダッシュボード側だけの拡張ロジック。
function monthsBetween(fromISO, toISO) {
  const a = new Date(fromISO + "T00:00:00");
  const b = new Date(toISO + "T00:00:00");
  const days = (b - a) / (1000 * 60 * 60 * 24);
  return days / 30.4375; // 平均月長で近似
}

function shiftedExpiry(item, params) {
  const shift = monthsBetween(BASE_AS_OF, params.asOfDate);
  return item.expiryM - shift;
}

// ---- 単一候補の因子計算 ---------------------------------------------------
function factors(item, params) {
  const lam = FIXED_PARAMS.lambdaRisk;
  const h = FIXED_PARAMS.leadTimeHMonths;

  const pinfo = effectiveP(item, params);
  const p = pinfo.p;

  const net = item.ret - item.cost;
  const Vraw = p * net * (1 - lam * (1 - p));

  const expiryM = shiftedExpiry(item, params);
  const d_i = expiryM - item.lead;

  let U;
  if (d_i < 0) {
    U = 0;
  } else {
    U = 1 / (1 + d_i / h);
  }

  const g = Math.min(Math.max(d_i, 0) / h, 1);
  const R = 1 - item.retSigma * g;

  return { net, Vraw, d_i, U, R, pEffective: p, ...pinfo };
}

// ---- 束ね（ベンダー単位バンドル）------------------------------------------
function buildBundles(activeContracts, params) {
  const beta = params.betaBundle;
  const byVendor = {};
  activeContracts.forEach((c) => {
    (byVendor[c.vendor] ??= []).push(c);
  });

  const bundles = [];
  for (const [vendor, members] of Object.entries(byVendor)) {
    if (members.length < 2) continue;
    const n = members.length;
    const ret = beta * members.reduce((s, m) => s + m.ret, 0);
    const cost = members.reduce((s, m) => s + m.cost, 0);
    let prodP = 1;
    members.forEach((m) => {
      prodP *= effectiveP(m, params).p;
    });
    const p = Math.pow(prodP, 1 / Math.sqrt(n));
    const retSigma = Math.max(...members.map((m) => m.retSigma));
    const expiryM = Math.min(...members.map((m) => shiftedExpiry(m, params)));
    const lead = Math.max(...members.map((m) => m.lead));

    bundles.push({
      id: `B-${vendor}`,
      name: `${vendor} 一括（${n}件）`,
      type: "Bundle",
      vendor,
      _shiftedExpiryM: expiryM, // すでにシフト済みなので factors() 内で二重シフトしない
      spend: members.reduce((s, m) => s + m.spend, 0),
      lead,
      cost,
      costSigma: 0,
      ret,
      retSigma,
      execProbability: p,
      contestProb: null,
      execProb: null,
      // 延長回数は構成契約の最大を引き継ぐ。1件でも延長常習なら
      // 束ねても現場任せでは動かないため、シグナルを消してはならない。
      extCount: Math.max(...members.map((m) => m.extCount || 0)),
      active: 1,
      isBundle: true,
      members: members.map((m) => m.id),
    });
  }
  return bundles;
}

// バンドル候補は expiryM をすでにシフト済みで保持しているため、
// factors() 呼び出し時に二重シフトしないよう専用パスを通す。
function factorsForBundle(bundle, params) {
  const lam = FIXED_PARAMS.lambdaRisk;
  const h = FIXED_PARAMS.leadTimeHMonths;
  const p = bundle.execProbability; // 束ねは合成済み確度をそのまま持つ
  const net = bundle.ret - bundle.cost;
  const Vraw = p * net * (1 - lam * (1 - p));
  const d_i = bundle._shiftedExpiryM - bundle.lead;
  let U = d_i < 0 ? 0 : 1 / (1 + d_i / h);
  const g = Math.min(Math.max(d_i, 0) / h, 1);
  const R = 1 - bundle.retSigma * g;
  const ext = bundle.extCount || 0;
  return {
    net, Vraw, d_i, U, R, pEffective: p,
    pContest: null, pExecEffective: null, extCount: ext,
    extEscalated: ext >= params.extThreshold, pSplitUsed: false,
  };
}

// ---- 判定ラベル ------------------------------------------------------------
function verdictOf(d_i, score, Vn, U, R, escalated) {
  if (d_i < 0) return { key: "失効", cls: "v-lapse" };
  if (score >= 0.25) return { key: "即着手", cls: "v-act" };
  // 繰り返し延長は、確度を割り引くのではなくここで浮上させる。
  // 「現場任せでは切り替わらない」という事実は、沈めるのではなく名指しする。
  if (escalated) return { key: "VMO主導候補", cls: "v-escalate" };
  if (U >= 0.5 && R < 0.7) return { key: "人間判断", cls: "v-judge" };
  if (Vn > 0.3 && U < 0.4) return { key: "温存", cls: "v-hold" };
  if (Vn < 0.1) return { key: "捨てる候補", cls: "v-drop" };
  return { key: "検討", cls: "v-check" };
}

// ---- メイン: 候補集合を構築してスコアリング --------------------------------
function scoreAll(params) {
  const active = CONTRACTS.filter((c) => Number(c.active) === 1);
  const excludedInactive = CONTRACTS.filter((c) => Number(c.active) !== 1).map((c) => c.name);

  let candidates = active.map((c) => ({ ...c, isBundle: false }));

  // 束ねは個別合計より有利かどうかに関わらず、常にランキングに候補として
  // 併記する（構成契約を除外・置換しない）。どちらを選ぶかはユーザーが
  // チェックボックスで手動選択する。
  if (params.useBundle) {
    const bundles = buildBundles(active, params);
    bundles.forEach((b) => candidates.push(b));
  }

  candidates = candidates.map((c) => {
    const f = c.isBundle ? factorsForBundle(c, params) : factors(c, params);
    return { ...c, ...f };
  });

  // 下限を定数で切らない（ゼロ除算だけ避ける）。src/modules/scoring.py と同基準。
  const maxV = Math.max(1e-9, ...candidates.map((c) => Math.max(c.Vraw, 0)));
  candidates.forEach((c) => {
    c.Vn = Math.max(c.Vraw, 0) / maxV;
    c.score = c.Vn * c.U * c.R;
    c.verdict = verdictOf(c.d_i, c.score, c.Vn, c.U, c.R, !!c.extEscalated);
  });
  candidates.sort((a, b) => b.score - a.score);

  return { candidates, excludedInactive };
}
